### PR TITLE
Stash Height Constraints

### PR DESCRIPTION
**Pull request by marqh**
I have moved the implied stash height constraints our of the PP rules and into the um_cf_map, so that they live alongside the other STASH interpretations

I have updated the translator tools to source this information from the metarelate metOcean knowledge base. 

As the extra information in the knowledge base causes some translations to be lost with the tool on master currently, I have held off the merge of the constraint knowledge awaiting confirmation that these Iris changes are acceptable
https://github.com/metarelate/metOcean/pull/48

### CODE REVIEW

**Comment by rhattersley** [comment]
> As the extra information in the knowledge base causes some translations to be lost with the tool on master currently, I have held off the merge of the constraint knowledge awaiting confirmation that these Iris changes are acceptable

It's not really clear from this what/where the problem is... but in this PR as it stands the changes within `lib/iris/...` look essentially OK. There are no lost translations - indeed there are two new ones.

I've not looked over the changes within `tools/...` in any detail, but clearly there are some very long lines so running pep8 on this code would help.

File: lib/iris/fileformats/um_cf_map.py [hunk 1]
```diff
@@ -815,6 +817,18 @@
     'm04s06i001': CFName('sea_surface_wind_wave_significant_height', None, 'm'),
     }
 
+_STASHCODE_IMPLIED_HEIGHTS = {
```
Comment (rhattersley) [inline lib/iris/fileformats/um_cf_map.py]: This should be a public name - i.e. ditch the leading underscore.

Comment (pelson) [inline lib/iris/fileformats/um_cf_map.py]: Do all of the "surface" type phenomenon deserve similar treatment to say that they have a height of 0? I don't propose removing the "surface" from the std name, just adding some extra metadata which should really be in the coordinate, not the phenomenon description...

**Comment by marqh** [comment]
Hi rhattersley.
> It's not really clear from this what/where the problem is

There's no problem for Iris here, this PR is to enhance the translation sharing metarelate can provide, particualrly with constrained STASH codes.  

Hopefully it gives metarelate benefit whilst limiting the added complexity in Iris

I've rebased to try and coax travis into life, it hasn't been playing nicely with this PR. I hope that hasn't added confusion

**Pull request merged**